File: plugin/src/Handlers/History.lua
local AuditLog = require(script.Parent.Parent.AuditLog)

local function historyHandler(payload)
    local action = tostring(payload.action or "list")
    if action == "list" then
        return { ok = true, data = { records = AuditLog.list() } }
    elseif action == "show" then
        if type(payload.commandId) ~= "string" or payload.commandId == "" then
            return { ok = false, error = "commandId missing" }
        end
        local record = AuditLog.show(payload.commandId)
        if not record then
            return { ok = false, error = "history record not found: " .. payload.commandId }
        end
        return { ok = true, data = record }
    elseif action == "undo" then
        if type(payload.commandId) ~= "string" or payload.commandId == "" then
            return { ok = false, error = "commandId missing" }
        end
        local result, err = AuditLog.undo(payload.commandId)
        if not result then
            return { ok = false, error = err }
        end
        return { ok = true, data = result }
    end
    return { ok = false, error = "unknown history action: " .. action }
end

return historyHandler

File: crates/rs/src/cli/history.rs
use crate::error::{AppError, AppResult};
use crate::protocol::messages::HistoryRequest;
use serde_json::Value;

pub fn list(port: u16, studio: Option<String>, json: bool) -> AppResult<()> {
    let response = request(port, studio, "list", None)?;
    if json {
        println!("{}", serde_json::to_string_pretty(&response)?);
    } else {
        let records = response
            .get("records")
            .and_then(Value::as_array)
            .cloned()
            .unwrap_or_default();
        if records.is_empty() {
            println!("No rs command history recorded in this Studio session.");
            return Ok(());
        }
        println!("Recent rs commands:");
        for record in records {
            let id = record
                .get("commandId")
                .and_then(Value::as_str)
                .unwrap_or("<unknown>");
            let kind = record
                .get("kind")
                .and_then(Value::as_str)
                .unwrap_or("<unknown>");
            let status = record
                .get("status")
                .and_then(Value::as_str)
                .unwrap_or("<unknown>");
            let at = record
                .get("startedAt")
                .and_then(Value::as_str)
                .unwrap_or("<unknown>");
            println!("  {id}  {kind}  {status}  {at}");
        }
    }
    Ok(())
}

pub fn show(port: u16, studio: Option<String>, id: String, json: bool) -> AppResult<()> {
    let response = request(port, studio, "show", Some(id))?;
    if json {
        println!("{}", serde_json::to_string_pretty(&response)?);
    } else {
        println!("{}", serde_json::to_string_pretty(&response)?);
    }
    Ok(())
}

pub fn undo(port: u16, studio: Option<String>, id: String, yes: bool, json: bool) -> AppResult<()> {
    if !yes {
        return Err(AppError::Other(
            "undo mutates Studio; pass --yes to restore a recorded rollback snapshot".into(),
        ));
    }
    let response = request(port, studio, "undo", Some(id))?;
    if json {
        println!("{}", serde_json::to_string_pretty(&response)?);
    } else {
        println!(
            "Undo {}",
            response["status"].as_str().unwrap_or("completed")
        );
        if let Some(path) = response.get("rootPath").and_then(Value::as_str) {
            println!("Restored: {path}");
        }
        if let Some(warnings) = response.get("warnings").and_then(Value::as_array) {
            for warning in warnings {
                println!("  - {}", warning.as_str().unwrap_or("<warning>"));
            }
        }
    }
    Ok(())
}

fn request(
    port: u16,
    studio: Option<String>,
    action: &str,
    command_id: Option<String>,
) -> AppResult<Value> {
    crate::cli::request::post(
        port,
        "history",
        "/history",
        &HistoryRequest {
            studio,
            action: action.to_string(),
            command_id,
        },
        150,
    )
}
